Ideals for Euclidean domains

wbhart opened the pull request:
Ideals:

1) In a Euclidean domain
2) For univariates over ZZ
3) For multivariates over ZZ

thofma commented:
Since it is WIP I have not looked at the details, but just noticed the use of `hasmethod`. Is this intended to stay? We usually just let the functions throw a `MethodError` everywhere else. Maybe we could do the same here?

Alternatively we could have a `gcdx(..) = error("Not implemented error")` fallback (similar to what @tthsqe12 and I discussed over at https://github.com/oscar-system/Oscar.jl/issues/790 for `divexact`).

wbhart commented:
The intention here was actually to deal with the case where a gcdx does not exist by using a different algorithm, rather than to raise a not implemented error. I realise that's not clear from the current code, and I'm not entirely sure how much I can do in a more general case. I will think about it before I remove the WIP.

wbhart commented:
I have now added ideal intersection. I have added everything I am going to here. If anyone else wants to add more functionality they can. But we should get this merged in the meantime.

There are two things blocking it from my point of view, which I will deal with early next week:

* The code currently raises an exception if xgcd is missing. I will remove those checks and just let Julia handle that.
* It seems isconstant is multiply defined (probably from a merge). I will fix this.

fingolfin commented on docs/src/ideal.md:
```diff
@@ -0,0 +1,180 @@
+array of generators and some additional fields for a parent object, etc. See
+the file `src/generic/GenericTypes.jl` for details.
+
+Parent objects of ideals have type `Generic.IdealSet{T}`.
```
So, just to understand: Is the parent of an ideal meant to stand for "the set of all ideals over rings with matching type"?

From a technical point of view: what are parents needed for / what could they be used for?

I note that ideals in Oscar don't seem to have parents (and that never bothered me so far, but then again maybe it's because I never had them, and so I never thought about it).

thofma replied:
Yes. We also have them for other ideal types in Hecke. They are useful/necessary when defining maps/morphisms where the domain/codomain consists of ideals.

fingolfin replied:
Aha, @fieker just told me that this can be useful to define homomorphisms from sets of ideals to other such sets. OK, makes sense, I guess

fingolfin commented on src/Poly.jl:
```diff
@@ -239,6 +239,14 @@ end
+@doc Markdown.doc"""
+    isconstant(a::PolynomialElem)
+
+Return `true` if the given polynomial is constant, i.e. has length less at most
```
Typo "less at most".

Also perhaps refer to `degree` (which every mathematician will understand) instead of `length`?

Indeed, I have to ask: what *is* the length of a polynomial? I couldn't find it documented in the manual, only in indirect references (e.g. the docstring for `degree` defines the degree via the length...). Looking at the code, I see what it means, but isn't this an implementation detail that should be hidden?

To add to my confusion, on  https://nemocas.github.io/AbstractAlgebra.jl/dev/mpolynomial/ it lists`length(f::MPolyElem)` in the APIs but without explanation. Looking at the docstring, it says: 
> Return the number of terms of the polynomial

but that of course is incompatible with the definition for univariate polynomials...?

fingolfin replied:
(sorry if `length` of an univariate polynomial is documented in the manual -- I honestly tried to find it, but probably my search skills were not good enough. If it is in there, I'd appreciate a pointer where to find it!)

wbhart replied:
It's defined on the interface page. The length function is not just left undefined for a univariate polynomial in the functionality section of the documentation but the length function is entirely undocumented there. It should be so thanks for picking that up.

I'll add a PR for that today if someone didn't already beat me to it.

wbhart replied:
isconstant was multiply defined. I removed the one with the incorrect docstring.

wbhart commented:
If the tests now pass, I am ready for this to be reviewed/merged.

Please note that I do not intend to add additional functionality or apply further speedups at this time. I believe it is important to merge this before it bit rots. I will return to it maybe next year to apply speedups (the most important of which would be tree pruning, which is not simple to add now).

wbhart commented:
I've now:

* Improved code coverage a bit (some comments above still apply)
* Added more Ideal constructors for good measure
* Fixed some doc issues
* Moved ideal tests to later in the test suite

I am pretty happy with it at this point.

thofma commented:
I am a bit puzzled by the normal form with respect to an ideal. What is the definition of this in the multivariate case? I can imagine what it is with respect to a list of polynomials, but what is it with respect to an ideal?

thofma commented:
I have two more general questions, but I don't think they need to be addressed in this PR here. We should get that in as fast as possible.

1. Another question would be the name, although that it is not too important as it can be changed later. `Generic.Ideal` is now taken a very general name with a very specific implementation and behaviour. 
2. It seems that generators are always "reduced" when constructing an ideal. Sometimes it is nice to just create ideals without doing actual computations (which may or not happen afterwards). What do you think of the idea of having `::gens` stored by default and another `::reduced_gens` field? Then we could also make `gens(Ideal(R, G)) == G`, which I think is always nice to have. It might also be helpful (notation-wise) if one wants to write an element of an ideal in terms of a given set of generators (which are not necessarily reduced). Finally, it would make them play nice with the other ideals in Oscar land.

Looking forward to your thoughts on these two matters.

wbhart commented:
Yeah I guess normal_form should be applied to some list of generators or something. I'm happy to change that.

I don't know how useful ideals are computationally if not reduced, but I'm ok with your number 1 and 2 if that's what is preferred.

wbhart commented:
The problem I have with letting normal_form take a list of generators is that it really only makes sense if that list of generators is a GB. But how do I check that without doing an expensive check? If I insist that ideals are always given as a GB then only allow taking normal_form wrt an ideal then the problem doesn't occur.

thofma commented:
Regarding the normal form, I think it is similar to `divrem(f::MPoly, Vector{MPoly})`? The `normal_form` function just returns something for which the leading term is not divisible by any leading term of any element in the list, right? The (strong) Gröbner basis property only comes into play to get uniqueness of the "remainder". Or probably the `normal_form` implementation here assumes that the second argument has the properties of a GB?

Regarding not always computing a GB on construction:
- Ideals are useful by just existing in the system and if it only means "I am storing an ideal here and not just a list of elements". (We had this a week ago in Oscar)
- `contains(I, J)` does not require a GB of `J`.
- (It is also not clear to me if the various operations between ideals benefit from first computing a Gröbner basis of the input ideals.)

wbhart commented:
I'm not aware of how to check containment for example without the strong reduction property. Without it the result depends on the order of the elements, the notion of division used and multiplication of the generators by units, even when the result would be zero if the generators were strongly reduced.

For example, some s-poly or g-poly could reduce the result further (even to zero). Or it can happen that reduction by the generators in one order can give zero whereas reduction in another order does not.

Similarly I don't know how to do intersection without reduction.

It's true you can give a list of generators for the union of ideals if you have a list of generators for each, without reduction. And some other very simple operations don't require reduction such as scalar multiplication. 

I guess if you want to support certain operations that can be computed this way without reduction then we have to go through all the rigmarole we have in Oscar of having separate objects for lists of generators and marking them as strongly reduced or not (or better yet having the equivalent of a Groebner basis *object* in addition to just a list of generators as has been proposed for Oscar itself, though I'm not sure what you'd call this object). But there are operations we can't support this way, especially in the Euclidean domain case.

Anyway, it's a lot of work on top of this PR. I don't see anything I could implement now without implementing that entire approach. The only thing that seems feasible to do before merging is rename Ideal to something else. But what I don't know. There are multiple different things here called ideal, so we probably need a number of different names.

wbhart commented:
Consider for example reduction of 6x^2 + 1 by the ideal generated by [9x^2 + 2, 6x^2 + 1]. Clearly the result will be zero if you reduce by the second element first but not if you reduce by the first element first. Note the leading term would be removed in either case if we were working over a field instead! So the situation is quite different in the case of ideals of a polynomial ring over a Euclidean domain instead of a field.

wbhart commented:
Though perhaps we should disambiguate the two multiple notions of reduction (strong and weak "reduction" of an ideal to a standard basis and reduction of a polynomial by a standard basis). It's unfortunate that I used reduce for the former and normal_form for the latter, whereas Singular uses std for the former and reduce for the latter.

But I'm just not sure what we should call an ideal that has been "reduced" to a standard basis. The literature seems to call steps in the algorithm reduction steps but then goes to some trouble to always refer to the result as a strong Groebner basis (though there are different notions of this depending on what properties one wants, I believe).

Actually I had thought standard form was the correct terminology here, so I am slightly surprised to see them referred to as Groebner bases in the literature.

I guess I am out of my depth on the naming and all the relevant notions and theorems.

wbhart commented:
Another question I have is, suppose we stop calling the ideal objects here Ideal. What would that be used for? Surely the notion of an ideal that contains an array of generators (and perhaps a flag indicating whether it is in `standard form') is about the most canonical notion one can have. Anything that has some other representation should surely be the thing that is renamed to something else.

Agreed, Oscar may not want to have an outer constructor called Ideal which by default performs reduction of the ideal upon construction. But there is no reason that the Ideal constructor in Oscar needs to call the Ideal constructor in AbstractAlgebra/Nemo.

It's true that we had a similar situation before with Module in AbstractAlgebra which ended up being renamed FPModule. I don't recall whether this was ultimately helpful or not.

fingolfin commented:
I think what @thofma was saying is that "`contains(I, J)` does not require a GB of `J`" -- i.e., it needs one for `I`, but not for `J`.

A natural way to resolve this is to compute the standard/Gröbner/whatever basis lazily. That works well in Oscar, at least. I see no fundamental problem doing that here, too? But that may of course simply be because I am not familiar with the code

wbhart commented:
Yes that's correct, contains(I, J) only needs a GB of I.

There's no fundamental problem computing the GB lazily here except that this may actually be (much) slower for some sequences of operations. If there was a theorem or even a heuristic that suggested otherwise, I wouldn't have any problem with it.

Anyway, I think the PR should be merged as is, and then when someone figures out what they want to do with it they can make the adjustments to their taste. I can't tell if this PR breaks Oscar due to the lift issue, but if not then I don't think merging as is will hurt anything. Oscar will probably use Singular rather than this code anyway so modifying it to match Oscar is rather a moot point.

I have a Number Theory application for the code as is if I ever get around to it. I had that sort of thing in mind more than a replacement for part of Singular.

fingolfin commented on docs/src/ideal.md:
```diff
@@ -0,0 +1,182 @@
+AbstractAlgebra.jl provides a module, implemented in `src/generic/Ideal.jl` for
+ideals of a Euclidean domain (assuming the existence of a `gcdx` function)
+or of a univariate or multivariate polynomial ring
+over the integers. Univariate and multivariate polynomial rings over other
```
Out of curiosity: the code seems to be mostly generic (judging by the type), so where does this restriction come in? Perhaps the use of `max` in the code is part of the reason? Or something else?

wbhart replied:
One needs heuristics for the "size" of polynomials. They are pretty subtle and depend very much on the ring. Over the integers things like the number of bits of the coefficients (appropriately weighted) are used.
